This Notebook implement the Polynomial regression of order 2 and 3 on the Insurance dataset where target variable is the charges incurred by various patients. Exhuastive EDA is already done in the Linear regression Notebook. where we found that smoker is the most important feature out of all 7 features.

Importing the necessary libraries

In [0]:
import pandas as pd
from sklearn import datasets
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import linear_model
from sklearn.preprocessing import  StandardScaler, LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score,mean_squared_error

In [13]:
from google.colab import files #Comment if not using Colab
uploaded = files.upload() #Comment if not using Colab

Saving insurance.csv to insurance (1).csv


In [5]:
data= pd.read_csv("insurance.csv") # Loading data 
data.head() # Checking first 5 records

(   age     sex     bmi  children smoker     region      charges
 0   19  female  27.900         0    yes  southwest  16884.92400
 1   18    male  33.770         1     no  southeast   1725.55230
 2   28    male  33.000         3     no  southeast   4449.46200
 3   33    male  22.705         0     no  northwest  21984.47061
 4   32    male  28.880         0     no  northwest   3866.85520,
 pandas.core.frame.DataFrame)

In [15]:
data.describe() #Checking summary statistics of the numerical features

age          sex  ...       region       charges
count  1338.000000  1338.000000  ...  1338.000000   1338.000000
mean     39.207025     0.505232  ...     1.515695  13270.422265
std      14.049960     0.500160  ...     1.104885  12110.011237
min      18.000000     0.000000  ...     0.000000   1121.873900
25%      27.000000     0.000000  ...     1.000000   4740.287150
50%      39.000000     1.000000  ...     2.000000   9382.033000
75%      51.000000     1.000000  ...     2.000000  16639.912515
max      64.000000     1.000000  ...     3.000000  63770.428010

[8 rows x 7 columns]

Encoding categorical data into numerical encoder for the easy computation.

In [0]:
#sex
le = LabelEncoder()
le.fit(data.sex.drop_duplicates()) 
data.sex = le.transform(data.sex)
# smoker or not
le.fit(data.smoker.drop_duplicates()) 
data.smoker = le.transform(data.smoker)
#region
le.fit(data.region.drop_duplicates()) 
data.region = le.transform(data.region)

In [17]:
data.corr()['charges'] # finding correaltion of target variable with input features

age         0.299008
sex         0.057292
bmi         0.198341
children    0.067998
smoker      0.787251
region     -0.006208
charges     1.000000
Name: charges, dtype: float64

Getting ready toi fit a plonomial regression model of order 2 and 3.

In [0]:
sc = StandardScaler()  #scaling data using standard scaling 
X_train = sc.fit_transform(X_train) # scaling input training features
X_test = sc.fit_transform(X_test) #scaling input test features

In [20]:
#Fitting degree 3 regressor.

X = data.drop(['charges','region'], axis = 1) # Input features 
Y = data.charges # output features

quad = PolynomialFeatures(degree = 3) # Creating an object of plonomial regressor
x_quad = quad.fit_transform(X) # Transforming the input data 

X_train,X_test,Y_train,Y_test = train_test_split(x_quad,Y, random_state = 0) #splitting the data

plr = LinearRegression().fit(X_train,Y_train) #fitting the model


Y_train_pred = plr.predict(X_train) # prediction on training set
Y_test_pred = plr.predict(X_test) # prediction on test set

print(plr.score(X_test,Y_test)) # scoring the model on holdout set

0.8810168906746892


In [22]:
#Fitting a degree 2 Regressor

X = data.drop(['charges','region'], axis = 1) # Input features
Y = data.charges # output features
 
quad = PolynomialFeatures(degree = 2) # Creating an object of plonomial regressor
x_quad = quad.fit_transform(X) # Transforming the input data

X_train,X_test,Y_train,Y_test = train_test_split(x_quad,Y, random_state = 0) #splitting the data

plr = LinearRegression().fit(X_train,Y_train) #fitting the model

Y_train_pred = plr.predict(X_train) # prediction on training set
Y_test_pred = plr.predict(X_test) # prediction on test set

print(plr.score(X_test,Y_test))  # scoring the model on holdout set

0.8849197344147228


Conclusion: The degree 2 regressor is better on this dataset and giving a accuracy of 88.5% on the holdout set. Which is even better than linear regression we trained and tested in earlier notebook.